File: Sprint4/F/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/F/
#include <iostream>
#include <algorithm>
#include <map>
#include <vector>

int main() {
    int n;
    std::cin >> n;
    std::map<std::string, std::vector<int>> map;

    for (int i = 0; i < n; ++i) {
        std::string str;
        std::cin >> str;
        std::sort(str.begin(), str.end());
        map[str].push_back(i);
    }

    std::vector<std::tuple<int, std::vector<int>>> result;
    result.reserve(map.size());
    for (const auto &[key, vec]: map) {
        result.emplace_back(vec.front(), vec);
    }

    std::sort(result.begin(), result.end());

    for (const auto &t: result) {
        for (int elem: std::get<1>(t)) {
            std::cout << elem << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}

File: Sprint4/F/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/F/

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let _: usize = input.trim().parse().unwrap();

    let mut map: std::collections::BTreeMap<String, Vec<usize>> = std::collections::BTreeMap::new();

    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    let words: Vec<String> = line
        .trim()
        .split_whitespace()
        .map(|x| x.to_string())
        .collect();

    for (index, word) in words.iter().enumerate() {
        let mut chars: Vec<char> = word.chars().collect();
        chars.sort();
        let key: String = chars.into_iter().collect();

        map.entry(key).or_default().push(index);
    }

    let mut result: Vec<(usize, Vec<usize>)> = Vec::new();
    for (_, vec) in &map {
        if !vec.is_empty() {
            result.push((vec[0], vec.clone()));
        }
    }

    result.sort();

    for t in &result {
        for &elem in &t.1 {
            print!("{} ", elem);
        }
        println!();
    }
}
